src/OpenLoco/Windows/Construction/Common.cpp

```diff
@@ -200,30 +201,192 @@ namespace OpenLoco::Ui::Windows::Construction
+        if (wnd == nullptr)
+        {
+            WindowManager::closeConstructionWindows();
+            Common::createConstructionWindow();
+        }
+        else
+        {
+            Common::resetWindow(*wnd, Common::widx::tab_construction);
+        }
+
+        _trackType = copyElement->trackObjectId();
+        _byte_1136063 = 0;
+        Common::setTrackOptions(_trackType);
+
+        _constructionHover = 0;
+        _byte_113607E = 1;
+        _trackCost = 0x80000000;
+        _byte_1136076 = 0;
+        _lastSelectedTrackModSection = 0;
+
+        Common::setNextAndPreviousTrackTile(*copyElement, pos);
+
+        const bool isCloserToNext = Common::isPointCloserToNextOrPreviousTile(Input::getDragLastLocation(), *viewport);
+
+        const auto chosenLoc = isCloserToNext ? *_nextTile : *_previousTile;
+        const auto chosenRotation = isCloserToNext ? _nextTileRotation : _previousTileRotation;
+        _x = chosenLoc.x;
+        _y = chosenLoc.y;
+        _word_1135FB8 = chosenLoc.z;
+        _constructionRotation = chosenRotation;
+        _lastSelectedTrackPiece = 0;
+        _lastSelectedTrackGradient = 0;
+
+        Common::refreshSignalList(_signalList, _trackType);
+        auto lastSignal = _scenarioSignals[_trackType];
+
+        if (lastSignal == 0xFF)
+            lastSignal = _signalList[0];
+
+        _lastSelectedSignal = lastSignal;
+
+        Common::refreshStationList(_stationList, _trackType, TransportMode::rail);
+
+        auto lastStation = _scenarioTrainStations[_trackType];
+
+        if (lastStation == 0xFF)
+            lastStation = _stationList[0];
+
+        _lastSelectedStationType = lastStation;
+
+        Common::refreshBridgeList(_bridgeList, _trackType, TransportMode::rail);
+
+        auto lastBridge = _scenarioBridges[_trackType];
+
+        if (lastBridge == 0xFF)
+            lastBridge = _bridgeList[0];
+
+        _lastSelectedBridge = lastBridge;
+
+        if (copyElement->has_4_80())
+        {
+            _lastSelectedBridge = copyElement->bridge();
+        }
+        Common::refreshModList(_modList, _trackType, TransportMode::rail);
+
+        _lastSelectedMods = copyElement->mods();
+        _byte_113603A = 0;
+        auto* window = WindowManager::find(WindowType::construction);
+
+        if (window != nullptr)
+        {
+            Common::setDisabledWidgets(window);
+        }
 
-        return reinterpret_cast<Window*>(regs.esi);
+        return window;
     }
 
     // 0x004A147F
-    Window* openAtRoad(Window* main, RoadElement* track, const Pos2 pos)
+    Window* openAtRoad(Window* main, RoadElement* road, const Pos2 pos)
     {
-        registers regs{};
-        regs.esi = X86Pointer(main);
-        regs.edx = X86Pointer(track);
-        regs.ax = pos.x;
-        regs.cx = pos.y;
-        call(0x004A147F, regs);
+        auto* viewport = main->viewports[0];
+        _backupTileElement = *reinterpret_cast<TileElement*>(road);
```

Comment:
copies the element as construction ghost clear could cause a reorg and make the pointer bad.